compiler/rustc_codegen_llvm/src/intrinsic_util.rs

```diff
@@ -0,0 +1,180 @@
+use rustc_codegen_ssa::mir::operand::OperandRef;
+use rustc_codegen_ssa::mir::operand::OperandValue;
+use rustc_codegen_ssa::mir::place::PlaceRef;
+use rustc_codegen_ssa::traits::{BaseTypeMethods, BuilderMethods, MiscMethods};
+use rustc_hir::LangItem;
+use rustc_middle::ty;
+use rustc_middle::ty::layout::LayoutOf;
+use rustc_middle::ty::subst::GenericArg;
+use rustc_middle::ty::subst::{Subst, SubstsRef};
+use rustc_middle::ty::util::IntTypeExt;
+use rustc_middle::ty::AdtDef;
+use rustc_span::DUMMY_SP;
+
+use crate::builder::Builder;
+use crate::value::Value;
+
+/// Recurse through the type of an rvalue and generate calls to
+/// `Collectable::set_collectable` for every component type which implements
+/// `Collectable`.
+///
+/// This is analogous to the MIR drop glue elaboration, except it must happen
+/// during codegen because it needs access to monomorphized types for the
+/// `ty.is_collectable` trait resolution.
+pub(crate) fn codegen_collectable_calls<'ll, 'tcx>(
+    builder: &mut Builder<'_, 'll, 'tcx>,
+    op: OperandRef<'tcx, &'ll Value>,
+    substs: SubstsRef<'tcx>,
+) {
+    let place = op.deref(builder.cx());
+    codegen_collectable_calls_for_value(builder, place, substs);
+}
+
+fn codegen_collectable_calls_for_value<'ll, 'tcx>(
+    builder: &mut Builder<'_, 'll, 'tcx>,
+    place: PlaceRef<'tcx, &'ll Value>,
+    substs: SubstsRef<'tcx>,
+) {
+    let ty = place.layout.ty.subst(builder.tcx, substs);
+    match ty.kind() {
+        ty::Infer(ty::FreshIntTy(_))
+            | ty::Infer(ty::FreshFloatTy(_))
+            | ty::Bool
+            | ty::Int(_)
+            | ty::Uint(_)
+            | ty::Float(_)
+            | ty::Never
+            | ty::FnDef(..)
+            | ty::FnPtr(_)
+            | ty::Char
+            | ty::GeneratorWitness(..)
+            // It's not possible to call `Collectable` trait methods behind a
```

Comment:
This says "it's not possible" but then seems to say "actually, you can do it, but it's not a good idea"?

Reply:
It means it's not possible for the compiler to do it. The user might have knowledge that means this is safe. I.e. your `Collectable` impl on `Box` can recurse into `Box.0` because the lifetime of the inner raw pointer is tied to the box, but the compiler can't know that.

Reply:
Do we maybe mean "It's not safe to call Collectable trait methods behind a ..."?

Reply:
So in cases like this, it's up to the user to call `<Box.0 as Collectable>::set_collectable` inside the `Box::set_collectable` trait method impl.

Reply:
I *think* I understand. Can we clarify the comment a bit?

Reply:
> Do we maybe mean "It's not safe to call Collectable trait methods behind a ..."?

I think our restriction is stronger than that: we're saying that it is _unsound_ for the compiler to call any `Collectable` trait method behind a raw pointer since it has no idea about the validity of that pointer. 

It is _unsafe_ for the user to do it as part of the `Collectable` trait method of an owning type.

What this comment is really saying is the compiler is going "whoah, raw pointer, i'm not touching this!"

Reply:
I'm with you! So at this point in the code what we mean is "we can't auto-insert a call with raw pointers, because we have no idea if the pointer is valid; if the user explicitly derefs a pointer and then does collectable stuff, that's fine."

Reply:
> I'm with you! So at this point in the code what we mean is "we can't auto-insert a call with raw pointers, because we have no idea if the pointer is valid; if the user explicitly derefs a pointer and then does collectable stuff, that's fine."

Yes! Exactly.

Reply:
Right. Can we mildly clarify things along the lines I said, so that idiots like me aren't confused in the future? :)

Reply:
How about something like this: eb8579a3aee3e6fb0e751d9a7234a5f8cbb5d8b